File: backend/modules/observability/domain/task/entity/filter.go
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0

package entity

// QueryType represents the operator applied to filter values.
type QueryType string

// QueryRelation represents the logical relation between multiple filter expressions.
type QueryRelation string

// FieldType describes the type of a field used in filter expressions.
type FieldType string

// TaskFieldName defines the supported task field names for filtering.
type TaskFieldName string

type TaskSourceValue string

const (
	QueryTypeMatch    QueryType = "match"
	QueryTypeEq       QueryType = "eq"
	QueryTypeNotEq    QueryType = "not_eq"
	QueryTypeLte      QueryType = "lte"
	QueryTypeGte      QueryType = "gte"
	QueryTypeLt       QueryType = "lt"
	QueryTypeGt       QueryType = "gt"
	QueryTypeExist    QueryType = "exist"
	QueryTypeNotExist QueryType = "not_exist"
	QueryTypeIn       QueryType = "in"
	QueryTypeNotIn    QueryType = "not_in"
	QueryTypeNotMatch QueryType = "not_match"

	QueryRelationAnd QueryRelation = "and"
	QueryRelationOr  QueryRelation = "or"

	FieldTypeString FieldType = "string"
	FieldTypeLong   FieldType = "long"
	FieldTypeDouble FieldType = "double"
	FieldTypeBool   FieldType = "bool"

	TaskFieldNameTaskStatus TaskFieldName = "task_status"
	TaskFieldNameTaskName   TaskFieldName = "task_name"
	TaskFieldNameTaskType   TaskFieldName = "task_type"
	TaskFieldNameSampleRate TaskFieldName = "sample_rate"
	TaskFieldNameCreatedBy  TaskFieldName = "created_by"
	TaskFieldNameTaskSource TaskFieldName = "task_source"

	TaskSourceUser TaskSourceValue = "user"
)

// TaskFilterFields aggregates multiple TaskFilterField expressions.
type TaskFilterFields struct {
	QueryAndOr   *QueryRelation
	FilterFields []*TaskFilterField
}

// GetQueryAndOr returns the relation between filter expressions.
func (f *TaskFilterFields) GetQueryAndOr() string {
	if f == nil || f.QueryAndOr == nil {
		return string(QueryRelationAnd)
	}
	return string(*f.QueryAndOr)
}

// TaskFilterField describes a single filter clause.
type TaskFilterField struct {
	FieldName  *TaskFieldName
	FieldType  *FieldType
	Values     []string
	QueryType  *QueryType
	QueryAndOr *QueryRelation
	SubFilter  *TaskFilterField
}

File: frontend/packages/loop-base/api-schema/src/api/idl/observability/domain/filter.ts
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0
import * as common from './common';
export { common };
export enum QueryType {
  Match = "match",
  Eq = "eq",
  NotEq = "not_eq",
  Lte = "lte",
  Gte = "gte",
  Lt = "lt",
  Gt = "gt",
  Exist = "exist",
  NotExist = "not_exist",
  In = "in",
  not_In = "not_in",
  NotMatch = "not_match",
}
export enum QueryRelation {
  And = "and",
  Or = "or",
}
export enum FieldType {
  String = "string",
  Long = "long",
  Double = "double",
  Bool = "bool",
}
export type TaskFieldName = string;
export const TaskFieldName_TaskStatus = "task_status";
export const TaskFieldName_TaskName = "task_name";
export const TaskFieldName_TaskType = "task_type";
export const TaskFieldName_SampleRate = "sample_rate";
export const TaskFieldName_CreatedBy = "created_by";
export interface FilterFields {
  query_and_or?: QueryRelation,
  filter_fields: FilterField[],
}
export interface FilterField {
  field_name?: string,
  field_type?: FieldType,
  values?: string[],
  query_type?: QueryType,
  query_and_or?: QueryRelation,
  sub_filter?: FilterFields,
  is_custom?: boolean,
  extra_info?: {
    [key: string | number]: string
  },
}
export interface FieldOptions {
  i64_list?: string[],
  f64_list?: number[],
  string_list?: string[],
}
export interface TaskFilterFields {
  query_and_or?: QueryRelation,
  filter_fields: TaskFilterField[],
}
export interface TaskFilterField {
  field_name?: TaskFieldName,
  field_type?: FieldType,
  values?: string[],
  query_type?: QueryType,
  query_and_or?: QueryRelation,
  sub_filter?: TaskFilterField,
}
export interface SpanFilterFields {
  /** Span 过滤条件 */
  filters?: FilterFields,
  /** 平台类型，不填默认是fornax */
  platform_type?: common.PlatformType,
  /** 查询的 span 标签页类型，不填默认是 root span */
  span_list_type?: common.SpanListType,
}